src/LanguageServer.ts

```diff
@@ -915,21 +931,134 @@ export class LanguageServer {
-        //WARNING: This only works for a few small xml cases because the vscode-brightscript-language extension
-        //already implemented this feature, and I haven't had time to port all of that functionality over to
-        //this codebase
-        //TODO implement for brs/bs also
+    public async onWorkspaceSymbol(params: WorkspaceSymbolParams) {
+        const results = util.flatMap(
+            await Promise.all(this.getWorkspaces().map(workspace => {
+                return workspace.builder.program.getWorkspaceSymbols();
+            })),
+            c => c
+        );
+
+        const allSymbols = Object.values(results.reduce((map, symbol) => {
+            const key = symbol.location.uri + symbol.name;
+            map[key] = symbol;
+            return map;
+        }, {}));
+        return allSymbols as SymbolInformation[];
+    }
+
+    public async onDocumentSymbol(params: DocumentSymbolParams) {
+        const pathAbsolute = util.uriToPath(params.textDocument.uri);
+        for (const workspace of this.getWorkspaces()) {
+            const file = workspace.builder.program.getFileByPathAbsolute(pathAbsolute);
+            if (file instanceof BrsFile) {
+                const symbols = await file.getDocumentSymbols();
+                return symbols;
+            }
+        }
+    }
+
+    private async onDefinition(params: TextDocumentPositionParams) {
         await this.waitAllProgramFirstRuns();
         let results = [] as Location[];
 
         let pathAbsolute = util.uriToPath(params.textDocument.uri);
         let workspaces = this.getWorkspaces();
         for (let workspace of workspaces) {
+            const locations = await workspace.builder.program.getDefinition(pathAbsolute, params.position);
             results = results.concat(
-                ...workspace.builder.program.getDefinition(pathAbsolute, params.position)
+                ...locations
             );
         }
+        // TODO I think this could be optimized better but it does the trick for now
+        const deduplicatedDefinitions = Object.values(results.reduce((map, location) => {
+            const range = location.range ?? Range.create(0, 0, 0, 0);
+            const start = range.start;
+            const end = range.end;
+            const key = `${location.uri}:${start.line}_${start.character}-${end.line}_${end.character}`;
+            map[key] = location;
+            return map;
+        }, {}));
+        return deduplicatedDefinitions as Location[];
+    }
+
+    private async onSignatureHelp(params: SignatureHelpParams) {
+        console.log('onSignatureHelp', 'params', params);
+
+        // Courtesy of George Cook :) He might call it whack but it works ¯\_(ツ)_/¯
+        //get the position of a symbol to our left
+        //1. get first bracket to our left, - then get the symbol before that..
+        //really crude crappy parser..
+        //TODO this is whack - it's not even LTR ugh..
```

Comment:
It's ok: simple and good enough. That was the basis of how I did code completion for a much more complex language 😁